prototypes/modified_zonal_approach/modified_zonal_approach.cc

```diff
@@ -0,0 +1,305 @@
+﻿/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019- by the Lethe authors
+ *
+ * This file is part of the Lethe library.
+ *
+ * The LEthe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Author: Bruno Blais, Polytechnique Montreal, 2019
+ */
+
+
+// This is a template folder for prototype executables to be created when
+// developing totally new Lethe functionnalities
+
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/timer.h>
+
+#include "mockup_systems.h"
+#include "write_distances.h"
+
+
+using namespace dealii;
+
+template <int dim>
+std::vector<double>
+calculate_distance(std::vector<Point<dim>> triangle,
+                   std::vector<Point<dim>> particles)
+{
+  std::vector<double> distances(particles.size());
+  const double        radius = 0.25;
+  auto               &p_0    = triangle[0];
+  auto               &p_1    = triangle[1];
+  auto               &p_2    = triangle[2];
+
+  const Tensor<1, dim> e_0 = p_1 - p_0;
+  const Tensor<1, dim> e_1 = p_2 - p_0;
+
+  const Tensor<1, dim> normal      = cross_product_3d(e_0, e_1);
+  const double         norm_normal = normal.norm();
+  const Tensor<1, dim> unit_normal = normal / norm_normal;
+
+  const double a   = e_0.norm_square();
+  const double b   = scalar_product(e_0, e_1);
+  const double c   = e_1.norm_square();
+  const double det = a * c - b * b;
+
+
+  // Pre-allocation for speed
+  Tensor<1, dim> vector_to_plane;
+  Point<dim>     pt_in_triangle;
+
+  unsigned int k = 0;
+  for (auto &part : particles)
+    {
+      vector_to_plane         = p_0 - part;
+      double distance_squared = scalar_product(vector_to_plane, unit_normal);
+
+      // If the particle is too far from the plane, set distance squared as an
+      // arbitrary distance and continue
+      if (distance_squared > (radius * radius))
+        {
+          distances[k] = std::sqrt(distance_squared);
+          ++k;
+          continue;
+        }
+
+      // Otherwise, do the full calculation taken from Eberly 2003
+      const double d = scalar_product(e_0, vector_to_plane);
+      const double e = scalar_product(e_1, vector_to_plane);
+
+      // Calculate necessary values;
+      double s = b * e - c * d;
+      double t = b * d - a * e;
+      // std::cout << "s " << s << " t " << t << std::endl;
```

Comment:
This comment shouldn't be removed?

Reply:
I think it's better to keep it for easy output.